File: LeetCodeCPP/872.leaf-similar-trees.cpp
/*
 * @lc app=leetcode id=872 lang=cpp
 *
 * [872] Leaf-Similar Trees
 */
#include <stack>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    bool leafSimilar(TreeNode *root1, TreeNode *root2)
    {
        stack<int> leaves1, leaves2;
        dfs(root1, leaves1);
        dfs(root2, leaves2);

        return leaves1 == leaves2;
    }

    void dfs(TreeNode *node, stack<int> &leaves)
    {
        if (node == nullptr)
            return;
        else if (node->left == nullptr && node->right == nullptr)
            leaves.push(node->val);
        else
        {
            dfs(node->left, leaves);
            dfs(node->right, leaves);
        }
    }
};
// @lc code=end

File: LeetCodePython/872.leaf-similar-trees.py
#
# @lc app=leetcode id=872 lang=python3
#
# [872] Leaf-Similar Trees
# ]

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        leaves1 = self.dfs(root1, [])
        leaves2 = self.dfs(root2, [])

        return leaves1 == leaves2

    def dfs(self, node: Optional[TreeNode], leaves: List[int]) -> List[int]:
        if node == None:
            return leaves

        if node.left == None and node.right == None:
            leaves.append(node.val)
        else:
            leaves = self.dfs(node.left, leaves)
            leaves = self.dfs(node.right, leaves)
        return leaves


# @lc code=end
